pandas/core/arrays/arrow/array.py

```diff
@@ -1541,21 +1513,94 @@ def pyarrow_meth(data, skip_nulls, **kwargs):
         if name == "median":
             # GH 52679: Use quantile instead of approximate_median; returns array
             result = result[0]
-        if pc.is_null(result).as_py():
-            return result
 
+        return self._maybe_cast_reduction_result(result, name)
+
+    def _values_for_reduction(self, name: str) -> pa.ChunkedArray:
+        """
+        Return the underlying ChunkedArray, possibly cast to a different type in
+        order to support reductions.
+
+        Parameters
+        ----------
+        name : str
+            Name of the function, supported values are:
+            { any, all, min, max, sum, mean, median, prod,
+            std, var, sem, kurt, skew }.
+
+        Returns
+        -------
+        pa.ChunkedArray
+        """
+        pa_type = self._pa_array.type
+
+        if name in ["any", "all"] and (
+            pa.types.is_integer(pa_type)
+            or pa.types.is_floating(pa_type)
+            or pa.types.is_duration(pa_type)
+            or pa.types.is_decimal(pa_type)
+        ):
+            # pyarrow only supports any/all for boolean dtype, we allow
+            #  for other dtypes, matching our non-pyarrow behavior
+            if pa.types.is_duration(pa_type):
+                data_to_cmp = self._pa_array.cast(pa.int64())
+            else:
+                data_to_cmp = self._pa_array
+
+            not_eq = pc.not_equal(data_to_cmp, 0)
+            data_to_reduce = not_eq
+
+        elif name in ["min", "max", "sum"] and pa.types.is_duration(pa_type):
+            data_to_reduce = self._pa_array.cast(pa.int64())
+
+        elif name in ["median", "mean", "std", "sem"] and pa.types.is_temporal(pa_type):
+            nbits = pa_type.bit_width
+            if nbits == 32:
+                data_to_reduce = self._pa_array.cast(pa.int32())
+            else:
+                data_to_reduce = self._pa_array.cast(pa.int64())
+
+        elif name in ["sem", "std", "var"] and pa.types.is_boolean(pa_type):
+            data_to_reduce = self._pa_array.cast(pa.uint8())
+
+        else:
+            data_to_reduce = self._pa_array
+
+        return data_to_reduce
+
+    def _maybe_cast_reduction_result(
```

Comment:
IIUC at least some of the bugfix is coming from adding this casting to _groupby_op.  Could split off from this PR one that implements _values_for_reduction and _maybe_cast_reduction_result?